x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java

```diff
@@ -232,19 +243,104 @@ private static void validateField(Map<?, ?> properties, String fieldName, boolea
         }
     }
 
-    private XContentBuilder resolveEnrichMapping(final EnrichPolicy policy) {
-        // Currently the only supported policy type is EnrichPolicy.MATCH_TYPE, which is a keyword type
-        final String keyType;
-        final CheckedFunction<XContentBuilder, XContentBuilder, IOException> matchFieldMapping;
+    private void resolveEnrichMapping(
+        final EnrichPolicy policy,
+        List<Map<String, Object>> mappings,
+        ActionListener<XContentBuilder> resultListener
+    ) {
         if (EnrichPolicy.MATCH_TYPE.equals(policy.getType())) {
-            matchFieldMapping = (builder) -> builder.field("type", "keyword").field("doc_values", false);
-            // No need to also configure index_options, because keyword type defaults to 'docs'.
+            createEnrichMappingBuilder((builder) -> builder.field("type", "keyword").field("doc_values", false), resultListener);
+        } else if (EnrichPolicy.RANGE_TYPE.equals(policy.getType())) {
+            resolveRangeTypeEnrichMapping(policy, mappings, resultListener);
         } else if (EnrichPolicy.GEO_MATCH_TYPE.equals(policy.getType())) {
-            matchFieldMapping = (builder) -> builder.field("type", "geo_shape");
+            createEnrichMappingBuilder((builder) -> builder.field("type", "geo_shape"), resultListener);
         } else {
-            throw new ElasticsearchException("Unrecognized enrich policy type [{}]", policy.getType());
+            resultListener.onFailure(new ElasticsearchException("Unrecognized enrich policy type [{}]", policy.getType()));
+        }
+    }
+
+    private void resolveRangeTypeEnrichMapping(
+        EnrichPolicy policy,
+        List<Map<String, Object>> mappings,
+        ActionListener<XContentBuilder> resultListener
+    ) {
+        String matchFieldPath = "properties." + policy.getMatchField().replace(".", ".properties.");
+        List<Map<String, String>> matchFieldMappings = mappings.stream()
+            .map(map -> ObjectPath.<Map<String, String>>eval(matchFieldPath, map))
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+
+        Set<String> types = matchFieldMappings.stream().map(map -> map.get("type")).collect(Collectors.toSet());
+
+        if (types.isEmpty()) {
```

Comment:
This is more of a style comment, but maybe let's do the error case in the else clause? I think it is easier to read, first the two valid cases and then everything else is an error case. Like this:

```java
if (types.isEmpty()) {
   // first valid case
} else if (types.size() == 1) {
   // second valid case
} else {
   // error case
}
```

Reply:
Considered the same thing 😅 but wanted to keep the delta small. I'll switch it around, also fiddled with a Optional but didn't seem to improve things, Kinda missing an Object to represent nothing/singular/multiple options.

Reply:
addressed in https://github.com/elastic/elasticsearch/pull/76110/commits/2924d2bb76c3cd027a1bf129f2de39b08a2461aa